Support for attachments

xiaoruoruo opened the pull request:
New features:

- Act as a file share target. Quickly capture images or files from other apps as a new note in orgzly.
- Add one or more attachments in the note editor, from "Attach file" menu.
- Render `attachment:` links, images can be rendered inline.

Three options for saving attachments:

- Option 1: Don't copy, just link to the file in the note, this is existing behavior
- Option 2: Copy file into a single directory specified by a pref
- Option 3: Copy file into a ID property based directory (org-attach style)

Future work:

- Delete attachments
- Insert a link into note for a given attachment
- Support for DropBoxRepo, WebDavRepo, GitRepo. Currently only DirectoryRepo is fully supported. Dropbox and Webdav have write support but not read.

Resolves
- https://github.com/orgzly/orgzly-android/issues/1
- https://github.com/orgzly/orgzly-android/issues/769

I have been using this myself locally for over a year and it works well for my use cases. Want to share with the revived orgzly community.

xiaoruoruo committed:
Support for attachments.

amberin reviewed (dismissed):
This looks fantastic to me. I suggest merging and creating a prerelease build ASAP, to allow for easy testing by as many users as possible.

cbkerr commented:
Hi @xiaoruoruo,

This would be an awesome addition! I agree it would be good to get a prerelease out soon.

Running in emulator (Pixel 7 API 33) to test, I can get Options 2 and 3 working.

Here is what happens with `Option 1: Don't copy, just link to the file in the note, this is existing behavior`. The following text gets put into the note content.
```text
content://com.google.android.apps.photos.contentprovider/-1/1/file%3A%2F%2F%2Fdata%2Fuser%2F0%2Fcom.google.android.apps.photos%2Fcache%2Fshare-cache%2Fmedia.tmp%3Ffilename%3D1280px-EmacsIcon.svg.png/ORIGINAL/NONE/image%2Fpng/1053993832

Cannot determine a local path to this file.
```

Running in an emulator, I was able to get `Option 2: Copy file into a single directory specified by a pref` and `Option 3: Copy file into a ID property based directory (org-attach style)` to add the correct link text to the note contents.

### Changes requested

1. The "Attach file" menu option only works if Option 3 is used. For options 1 and 2, the attachment is added in the list view, but when saving and re-opening the note, it does not persist.
2. Consider renaming "Attach file" to something that can encompass "Link file" "Copy file" or "Attach file" and make it clear the behavior depends on the option.
3. I think new behavior should be:
    - take whatever action the user has configured (link, copy, or attach)
    - add the appropriate link text to the note at the location of the cursor (`[[file:image.png]]` for Option 1 and 2 and `[[attachment:image.png]]` for Option 3)
4. There is an extra divider liner appears even on notes that don't have attachments
    <img width="455" alt="image" src="https://github.com/orgzly-revived/orgzly-android-revived/assets/21282461/ec8d9714-5b12-4070-823d-ec7c4165aad3">

cbkerr commented:
and @amberin I feel like these would be good things to address before a prerelease build

cbkerr commented:
>Consider renaming "Attach file" to something that can encompass "Link file" "Copy file" or "Attach file" and make it clear the behavior depends on the option.
I think new behavior should be:
 >   take whatever action the user has configured (link, copy, or attach)


I think I came up with a better idea. There could be submenus like how the "Clock" menu works now.

It could be called: "Insert file..." that goes to 3 actions: "Link file", "Copy file", and "Org-attach file"

Then go on to this like before:
 >   add the appropriate link text to the note at the location of the cursor ([[file:image.png]] for Option 1 and 2 and [[attachment:image.png]] for Option 3)

amberin commented:
@Wjxfi Sorry, if there were any updates, you would see them in this PR. It's pretty clear where we are: @xiaoruoruo suggested code changes, @cbkerr pointed out some issues that need to be resolved, and nothing has happened since then. So, apparently, nobody has yet had the time to do the remaining work.

superhans2 commented:
I've checked out the commit. Seems broken? 

For some reason (maybe unrelated to this commit idk, or just an older bug) the org files that sync with the directory repository are empty (even though the "notebooks" in the app will have content). Makes it harder to tell what is going on. 

Option 1 - linking - doesn't work

Option 2 - "copy to the attachment directory root - I can't seem to get working, it just uses the repository root to store attachments. 

Option 3 - "copy to ID property based directory under attachment directory" - does work, and uses the attachment directory I've specified in the attachment directory setting.

options 2 & 3 actually do copy the files into the repository (3 doing so as expected) but I don't know how the documents are embedded in the org file, if its doing this correctly because of the bug where I can't see the content of the sync-ed org files.

xiaoruoruo committed:
Merge branch 'master' into addimage1, resolving conflicts

xiaoruoruo committed:
The "Attach file" now has a submenu containing:

- Link file: Inserts a link ([[file:path]]).
- Copy file: Copies file to default directory and inserts link ([[file:filename]]).
- Org-attach file: Copies file to ID-based directory and inserts attachment link ([[attachment:filename]]).

xiaoruoruo committed:
Merged with upstream.

xiaoruoruo committed:
Fix divider visibility when no attachments.

xiaoruoruo committed:
Refine attachment handling, update attachment tests.

xiaoruoruo committed:
Merge remote-tracking branch 'upstream/master' into addimage1

xiaoruoruo commented:
Tests are updated, they are passing for me locally. This is ready for review.

amberin commented:
> Tests are updated, they are passing for me locally.

Is it possible that your local device had some repository manually configured? Because one test is failing with a "no repositories configured" error.

xiaoruoruo committed:
Refactor tests.

xiaoruoruo committed:
Merge branch 'master' into addimage1

xiaoruoruo committed:
Verify the intent in the test.

xiaoruoruo commented on app/src/androidTest/java/com/orgzly/android/repos/DirectoryRepoTest.java:
Done. Attachment UI tests moved to DocumentRepo instead of DirectoryRepo. I had to create TestDocumentsProvider to remove the dependency on flaky system picker automation.

xiaoruoruo commented:
Please take another look on the updated tests.

amberin commented:
SEVERE: Execute com.orgzly.android.espresso.OrgProtocolTest.testOrgProtocolOpensNote: FAILED
java.io.IOException: java.lang.Error: java.io.IOException: No repositories configured
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1173)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:919)
Caused by: java.io.IOException: No repositories configured
at com.orgzly.android.data.DataRepository.defaultRepoForSavingBook(DataRepository.kt:207)
at com.orgzly.android.data.DataRepository.listFiles(DataRepository.kt:1801)
at com.orgzly.android.ui.note.NoteViewModel.loadAttachments(NoteViewModel.kt:126)
at com.orgzly.android.ui.note.NoteViewModel.loadData$lambda$0(NoteViewModel.kt:104)

xiaoruoruo committed:
Merge master

xiaoruoruo committed:
Fix OrgProtocolTest: add repository setup to handle eager attachment loading

xiaoruoruo commented:
Thanks for the catch, please take another look.

amberin commented:
The paths in com.orgzly.android.repos.DocumentRepoTest do not match the expected ones. What might have changed the paths?

xiaoruoruo committed:
Fix DocumentRepo test failures caused by invalid SAF URI construction

- Replaced manual string concatenation in DocumentRepo with DocumentFile.findFile() to correctly resolve nested files and directories.
- Updated SyncRepoTest.kt with dynamic URI prefix resolution for cross-provider compatibility.

xiaoruoruo committed:
Fix DropboxRepoTest.

xiaoruoruo commented:
Fixed that test.

amberin commented on app/src/main/java/com/orgzly/android/BookName.java:
Please explain, in the commit message, which problem this change is intended to solve.

amberin commented on app/src/main/java/com/orgzly/android/repos/DocumentRepo.java:
What problem are we solving here? These are *very* consequential changes, so they need to be very clearly motivated.

We do have well-known issues with some ContentProviders and the handling of subdirectories (see primarily #381 and #554). But this is way too important code to just throw in some fixes on the side. Those issues definitely need to be handled in a separate PR, which looks at this from more angles than just making some tests pass.

amberin reviewed (changes requested):
Please remove all the .txt and .md files, or explain why you think they should be added to the repository.
